[1.12.2] Problem of incompatibility with another mod

DELGADO VELASQUEZ JAVIER ALEXANDER created an issue 2021-05-29:
> Good morning, this mod in its versions for 1.12.2 contains an incompatibility with the mod "Astral Sorcery" version 1.12.2 -1.10.20, the problem is that it simply prevents the ability to fire both firearms and cannons, it does not show crash, but it makes what you want to achieve with the mod useless.
> 
> I would appreciate your help. And sorry for my English, it is not my mother tongue.
> 
> Note: this is my list of mods and only by removing Astral Soercery can I use your mod. 10 attempts to remove and put to confirm. I also attach the confictive mod and its dependency "the dependency alone doesn't seem to affect", I don't know if it's really the right thing to do but I'm not very expert in putting this type of post.
> 
> 
> [1.12.2]DamageIndicatorsMod-3.5.1.jar"
> AbyssalCraft-1.12.2-1.9.12.jar"
> AmbientSounds_v3.0.10_mc1.12.2.jar"
> Animalium-0.3.8.jar"
> Astralsorcery-1.12.2-1.10.20.jar"
> backpack-3.0.2-1.12.2.jar"
> Baubles-1.12-1.5.2.jar"
> BiblioCraft[v2.4.5][MC1.12.2].jar"
> BloodMagic-1.12.2-2.4.2-104.jar"
> candyworld-1.1.5.jar"
> ChestTransporter-1.12.2-2.8.8.jar"
> Clans-1.12.2-2.1.3.jar"
> Comes-Alive-espanol.jar
> CreativeCore_v1.10.3_mc1.12.2.jar
> CustomNPCs_1.12.2-(01Oct19).jar
> ElectroblobsWizardry-4.2.10-MC1.12.2 (1).jar
> Explorercraft-2.2.9.jar"
> Forgelin-1.8.4.jar"
> future-mc-1.12.2-0.2.4.1.jar"
> GottschCore-mc1.12.2-f14.23.5.2854-v1.14.0.jar"
> GrimoireOfGaia3-1.12.2-1.7.2.jar
> Guide-API-1.12-2.1.8-63.jar"
> iceandfire-1.9.1-1.12.2.jar"
> InfernalMobs-1.12.2.jar"
> jei_1.12.2-4.15.0.291.jar"
> journeymap-1.12.2-5.5.5.jar"
> llibrary-1.7.19-1.12.2.jar"
> magicalcrops-1.0.2-MC-1.12.jar"
> Metallurgy-4-Reforged-0.2.3.jar"
> mowziesmobs-1.5.4.jar"
> oldguns-1.12.2-2.0.4.jar"
> OptiFine_1.12.2_HD_U_G5.jar"
> Overlord-1.12.2-2.4.14.jar"
> Overlord-1.12.2-2.4.14-sources.jar"
> Pam's+HarvestCraft+1.12.2zg.jar
> RecurrentComplex-1.4.8.2.jar
> RoguelikeDungeons-1.12.2-1.8.0.jar
> rustic-1.1.1.jar
> SanLib-1.12.2-1.5.1.jar
> SpartanShields-1.12.2-1.5.4.jar
> SpartanWeaponry-1.12.2-beta-1.3.8.jar
> SRParasites-1.12.2v1.8.2.jar
> Thaumcraft-1.12.2-6.1.BETA26.jar
> tombstone-3.5.9-1.12.2.jar
> "Treasure2-mc1.12.2-f14.23.5.2854-v1.16.1.jar
> VariedCommodities_1.12.2(15may18.jar
> "Corldedit-forge-mc1.12-6.1.8-dist.jar
> 
> 
> On my server, the console when firing a cannon throws this message.
> 
> 
> [02:01:56] [Server thread/ERROR] [FML]: Exception caught during firing event net.minecraftforge.event.entity.EntityJoinWorldEvent@4a35a64c:
> Cliente> java.lang.NullPointerException: null
> Cliente> at net.minecraft.network.datasync.EntityDataManager.func_187225_a(EntityDataManager.java:150) ~[nb.class:?]
> Cliente> at net.minecraft.entity.projectile.EntityArrow.func_70241_g(EntityArrow.java:636) ~[aeh.class:?]
> Cliente> at hellfirepvp.astralsorcery.common.constellation.perk.attribute.type.AttributeCritMultiplier.onArrowCt(AttributeCritMultiplier.java:42) ~[AttributeCritMultiplier.class:?]
> Cliente> at net.minecraftforge.fml.common.eventhandler.ASMEventHandler_47_AttributeCritMultiplier_onArrowCt_EntityJoinWorldEvent.invoke(.dynamic) ~[?:?]
> Cliente> at net.minecraftforge.fml.common.eventhandler.ASMEventHandler.invoke(ASMEventHandler.java:90) ~[ASMEventHandler.class:?]
> Cliente> at net.minecraftforge.fml.common.eventhandler.EventBus.post(EventBus.java:182) ~[EventBus.class:?]
> Cliente> at net.minecraft.world.World.func_72838_d(World.java:1209) ~[amu.class:?]
> Cliente> at net.minecraft.world.WorldServer.func_72838_d(WorldServer.java:1058) ~[oo.class:?]
> Cliente> at zach2039.oldguns.common.tile.TileEntityStationaryArtillery.lambda$processPlayerInteraction$0(TileEntityStationaryArtillery.java:512) ~[TileEntityStationaryArtillery.class:?]
> Cliente> at java.util.ArrayList.forEach(Unknown Source) [?:1.8.0_291]
> Cliente> at zach2039.oldguns.common.tile.TileEntityStationaryArtillery.processPlayerInteraction(TileEntityStationaryArtillery.java:502) [TileEntityStationaryArtillery.class:?]
> Cliente> at zach2039.oldguns.common.block.BlockStationaryCannon.func_180639_a(BlockStationaryCannon.java:164) [BlockStationaryCannon.class:?]
> Cliente> at net.minecraft.server.management.PlayerInteractionManager.func_187251_a(PlayerInteractionManager.java:449) [or.class:?]
> Cliente> at net.minecraft.network.NetHandlerPlayServer.func_184337_a(NetHandlerPlayServer.java:741) [pa.class:?]
> Cliente> at net.minecraft.network.play.client.CPacketPlayerTryUseItemOnBlock.func_148833_a(SourceFile:55) [ma.class:?]
> Cliente> at net.minecraft.network.play.client.CPacketPlayerTryUseItemOnBlock.func_148833_a(SourceFile:11) [ma.class:?]
> Cliente> at net.minecraft.network.PacketThreadUtil$1.run(SourceFile:13) [hv$1.class:?]
> Cliente> at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:1.8.0_291]
> Cliente> at java.util.concurrent.FutureTask.run(Unknown Source) [?:1.8.0_291]
> Cliente> at net.minecraft.util.Util.func_181617_a(SourceFile:46) [h.class:?]
> Cliente> at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:723) [MinecraftServer.class:?]
> Cliente> at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:668) [MinecraftServer.class:?]
> Cliente> at net.minecraft.server.integrated.IntegratedServer.func_71217_p(IntegratedServer.java:185) [chd.class:?]
> Cliente> at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:526) [MinecraftServer.class:?]
> Cliente> at java.lang.Thread.run(Unknown Source) [?:1.8.0_291]
> Cliente> [02:01:56] [Server thread/ERROR] [FML]: Index: 10 Listeners:
> Cliente> [02:01:56] [Server thread/ERROR] [FML]: 0: HIGHEST
> Cliente> [02:01:56] [Server thread/ERROR] [FML]: 1: ASM: net.minecraftforge.common.ForgeInternalHandler@3949fcdc onEntityJoinWorld(Lnet/minecraftforge/event/entity/EntityJoinWorldEvent;)V
> Cliente> [02:01:56] [Server thread/ERROR] [FML]: 2: ASM: hellfirepvp.astralsorcery.common.util.BlockDropCaptureAssist@27575e55 onDrop(Lnet/minecraftforge/event/entity/EntityJoinWorldEvent;)V
> Cliente> [02:01:56] [Server thread/ERROR] [FML]: 3: HIGH
> Cliente> [02:01:56] [Server thread/ERROR] [FML]: 4: ASM: hellfirepvp.astralsorcery.common.constellation.perk.attribute.type.AttributeCritChance@dc154bd7 onArrowCt(Lnet/minecraftforge/event/entity/EntityJoinWorldEvent;)V
> Cliente> [02:01:56] [Server thread/ERROR] [FML]: 5: NORMAL
> Cliente> [02:01:56] [Server thread/ERROR] [FML]: 6: ASM: baubles.common.event.EventHandlerEntity@357e6334 playerJoin(Lnet/minecraftforge/event/entity/EntityJoinWorldEvent;)V
> Cliente> [02:01:56] [Server thread/ERROR] [FML]: 7: ASM: hellfirepvp.astralsorcery.common.constellation.perk.attribute.type.AttributeArrowSpeed@b96d0033 onArrowFire(Lnet/minecraftforge/event/entity/EntityJoinWorldEvent;)V
> Cliente> [02:01:56] [Server thread/ERROR] [FML]: 8: ASM: hellfirepvp.astralsorcery.common.event.listener.EventHandlerEntity@2d8853ee onSpawnDropCloud(Lnet/minecraftforge/event/entity/EntityJoinWorldEvent;)V
> Cliente> [02:01:56] [Server thread/ERROR] [FML]: 9: LOWEST
> Cliente> [02:01:56] [Server thread/ERROR] [FML]: 10: ASM: hellfirepvp.astralsorcery.common.constellation.perk.attribute.type.AttributeCritMultiplier@8bcd7c54 onArrowCt(Lnet/minecraftforge/event/entity/EntityJoinWorldEvent;)V
> Cliente> [02:01:56] [Server thread/FATAL] [minecraft/MinecraftServer]: Error executing task
> Cliente> java.util.concurrent.ExecutionException: java.lang.NullPointerException
> Cliente> at java.util.concurrent.FutureTask.report(Unknown Source) ~[?:1.8.0_291]
> Cliente> at java.util.concurrent.FutureTask.get(Unknown Source) ~[?:1.8.0_291]
> Cliente> at net.minecraft.util.Util.func_181617_a(SourceFile:47) [h.class:?]
> Cliente> at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:723) [MinecraftServer.class:?]
> Cliente> at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:668) [MinecraftServer.class:?]
> Cliente> at net.minecraft.server.integrated.IntegratedServer.func_71217_p(IntegratedServer.java:185) [chd.class:?]
> Cliente> at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:526) [MinecraftServer.class:?]
> Cliente> at java.lang.Thread.run(Unknown Source) [?:1.8.0_291]
> Cliente> Caused by: java.lang.NullPointerException
> Cliente> at net.minecraft.network.datasync.EntityDataManager.func_187225_a(EntityDataManager.java:150) ~[nb.class:?]
> Cliente> at net.minecraft.entity.projectile.EntityArrow.func_70241_g(EntityArrow.java:636) ~[aeh.class:?]
> Cliente> at hellfirepvp.astralsorcery.common.constellation.perk.attribute.type.AttributeCritMultiplier.onArrowCt(AttributeCritMultiplier.java:42) ~[AttributeCritMultiplier.class:?]
> Cliente> at net.minecraftforge.fml.common.eventhandler.ASMEventHandler_47_AttributeCritMultiplier_onArrowCt_EntityJoinWorldEvent.invoke(.dynamic) ~[?:?]
> Cliente> at net.minecraftforge.fml.common.eventhandler.ASMEventHandler.invoke(ASMEventHandler.java:90) ~[ASMEventHandler.class:?]
> Cliente> at net.minecraftforge.fml.common.eventhandler.EventBus.post(EventBus.java:182) ~[EventBus.class:?]
> Cliente> at net.minecraft.world.World.func_72838_d(World.java:1209) ~[amu.class:?]
> Cliente> at net.minecraft.world.WorldServer.func_72838_d(WorldServer.java:1058) ~[oo.class:?]
> Cliente> at zach2039.oldguns.common.tile.TileEntityStationaryArtillery.lambda$processPlayerInteraction$0(TileEntityStationaryArtillery.java:512) ~[TileEntityStationaryArtillery.class:?]
> Cliente> at java.util.ArrayList.forEach(Unknown Source) ~[?:1.8.0_291]
> Cliente> at zach2039.oldguns.common.tile.TileEntityStationaryArtillery.processPlayerInteraction(TileEntityStationaryArtillery.java:502) ~[TileEntityStationaryArtillery.class:?]
> Cliente> at zach2039.oldguns.common.block.BlockStationaryCannon.func_180639_a(BlockStationaryCannon.java:164) ~[BlockStationaryCannon.class:?]
> Cliente> at net.minecraft.server.management.PlayerInteractionManager.func_187251_a(PlayerInteractionManager.java:449) ~[or.class:?]
> Cliente> at net.minecraft.network.NetHandlerPlayServer.func_184337_a(NetHandlerPlayServer.java:741) ~[pa.class:?]
> Cliente> at net.minecraft.network.play.client.CPacketPlayerTryUseItemOnBlock.func_148833_a(SourceFile:55) ~[ma.class:?]
> Cliente> at net.minecraft.network.play.client.CPacketPlayerTryUseItemOnBlock.func_148833_a(SourceFile:11) ~[ma.class:?]
> Cliente> at net.minecraft.network.PacketThreadUtil$1.run(SourceFile:13) ~[hv$1.class:?]
> Cliente> at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) ~[?:1.8.0_291]
> Cliente> at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:1.8.0_291]
> Cliente> at net.minecraft.util.Util.func_181617_a(SourceFile:46) ~[h.class:?]
> Cliente> ... 5 more
>

--- a/src/main/java/zach2039/oldguns/common/entity/EntityProjectile.java
+++ b/src/main/java/zach2039/oldguns/common/entity/EntityProjectile.java
@@ -127,6 +127,9 @@ public class EntityProjectile extends EntityArrow
 		
 		/* Register projectile size for rendering. */
 		dataManager.register(PROJECTILE_SIZE, 1f);
+		
+		/* Register CRITICAL flag for critical damage application. */
+		dataManager.register(CRITICAL, (byte)0);
 	}
 	
 	/**

--- a/src/main/java/zach2039/oldguns/common/entity/EntityProjectile.java
+++ b/src/main/java/zach2039/oldguns/common/entity/EntityProjectile.java
@@ -60,6 +60,7 @@ public class EntityProjectile extends EntityArrow
 	private static final DataParameter<Float> EFFECTIVE_RANGE = EntityDataManager.<Float>createKey(EntityProjectile.class, DataSerializers.FLOAT);
 	private static final DataParameter<String> SHOOTING_ENTITY = EntityDataManager.<String>createKey(EntityProjectile.class, DataSerializers.STRING);
 	private static final DataParameter<Float> DAMAGE = EntityDataManager.<Float>createKey(EntityProjectile.class, DataSerializers.FLOAT);
+	private static final DataParameter<Byte> CRITICAL = EntityDataManager.<Byte>createKey(EntityProjectile.class, DataSerializers.BYTE);
 	
 	protected int xTile = -1;
 	protected int yTile = -1;
@@ -102,6 +103,13 @@ public class EntityProjectile extends EntityArrow
 		this.setShootingEntity(shooter);
 	}
 
+	@Override
+	public boolean getIsCritical()
+    {
+        byte b0 = ((Byte)this.dataManager.get(CRITICAL)).byteValue();
+        return (b0 & 1) != 0;
+    }
+	
 	@Override
 	public void entityInit()
 	{		

